File: Python3/457. Circular Array Loop.py
# 44ms, 60.92%; 14MB, 100.00%
class Solution:
    def circularArrayLoop(self, nums: list) -> bool:
        is_visited = [False] * len(nums)
        for i in range(len(nums)):
            if not is_visited[i]:
                is_visited[i] = True
                if self.helper(nums, i, is_visited): return True
        return False

    def helper(self, nums: list, beg: int, is_visited: list) -> bool:
        length = len(nums)
        i, j, vals = beg, (beg + nums[beg]) % length, set([beg])
        while True:
            if i == j or nums[i] * nums[j] < 0: return False
            if is_visited[j] and j in vals: return True
            is_visited[j] = True
            vals.add(j)
            i, j = j, (j + nums[j]) % length
        return False

File: 401 - 600/457. Circular Array Loop.cc
// 0ms, 100.00%; 8.7MB, 10.00%
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool circularArrayLoop(const vector<int> &nums) {
        int size = nums.size();
        vector<bool> isVisited(size, false);
        for (int i = 0; i < size; ++i) {
            if (!isVisited[i]) {
                isVisited[i] = true;
                if (helper(nums, i, isVisited)) return true;
            }
        }
        return false;
    }

private:
    bool helper(const vector<int> &nums, int i, vector<bool> &isVisited) {
        int size = nums.size();
        int j = (i + nums[i]) % size;
        if (j < 0) j += size;
        unordered_set<int> tmp;
        tmp.insert(i);
        while (true) {
            if (i == j || nums[i] * nums[j] < 0) return false;
            if (isVisited[j] && tmp.find(j) != tmp.end()) return true;
            isVisited[j] = true;
            tmp.insert(j);
            i = j, j = (j + nums[j]) % size;
            if (j < 0) j += size;
        }
        return false;
    }
};
